scripts/research/lib/batch02_evidence_retention.py

```
@@ -0,0 +1,1036 @@
+    result_path: Path,
+    expected_sha256: str,
+    run_identity_sha256: str,
+    code_freeze: VerifiedCodeFreeze,
+) -> DurableArchiveReceipt:
+    """Archive exact persisted bytes and independently verify remote identity."""
+    reservation.assert_minted()
+    backend = _BACKEND_BY_ID[id(reservation)]
+    try:
+        freeze = verify_git_freeze(code_freeze.repo_root, code_freeze.code_sha)
+    except Exception as exc:
+        raise PreOutcomeRetentionError("execution worktree is no longer frozen") from exc
```

Ruff (0.16.3)

[warning] Avoid specifying long messages outside the exception class

(TRY003)